### DISCUSSION TITLE
Self-hosting Requal (Requal server installation)

### DISCUSSION THREAD

**Discussion by hlageek** [RE-QDA/requal#148] [Show and tell]
Tips for self-hosted Requal server installation.

**Comment by martin-hajek** [comment]
### **My installation of Requal on Ubuntu Server 24.04** (running as a virtual machine on my laptop)
<br/>

**Server requirements:** 
Ubuntu server 24.04 installation requirements
Minimum 10 GB, preferably 20 GB of free space 
<br/>
**1. Create a user _shiny_ and add her to the _sudo_ group**
(After the installation, the shiny user will be removed from this group again.)

`admin@virtual-server:~$ sudo adduser shiny`

`admin@virtual-server:~$ sudo adduser shiny sudo`

Switch the user to shiny

`admin@virtual-server:~$ su - shiny`

<br/>

**2. Install additional linux packages** 

`shiny@virtual-server:~$ sudo apt install make build-essential libz-dev libpq-dev`

<br/>

**3. Install PostgreSQL** (https://documentation.ubuntu.com/server/how-to/databases/install-postgresql/)

`shiny@virtual-server:~$ sudo apt install postgresql`

`shiny@virtual-server:~$ sudo systemctl enable postgresql`

In postgres, create a _shiny_ user and a _requal_ database

`shiny@virtual-server:~$ sudo -u postgres psql`
<br/>

In the database prompt 

`postgres=# CREATE USER shiny WITH PASSWORD 'password';`

`postgres=# ALTER USER shiny WITH SUPERUSER;`

`postgres=# CREATE DATABASE requal;`

`postgres=# GRANT ALL PRIVILEGES ON DATABASE requal TO shiny;`

`postgres=# \q`

<br/>

**4. Install R** (https://cloud.r-project.org/)

`shiny@virtual-server:~$ sudo apt update -qq`

`shiny@virtual-server:~$ sudo apt install --no-install-recommends software-properties-common dirmngr`

`shiny@virtual-server:~$ wget -qO- https://cloud.r-project.org/bin/linux/ubuntu/marutter_pubkey.asc | shiny@virtual-server:~$ sudo tee -a /etc/apt/trusted.gpg.d/cran_ubuntu_key.asc`

`shiny@virtual-server:~$ sudo add-apt-repository "deb https://cloud.r-project.org/bin/linux/ubuntu $(lsb_release -cs)-cran40/"`

`shiny@virtual-server:~$ sudo apt install --no-install-recommends r-base`

<br/>

**5. Create a directory /srv/shiny-server/ and make it the home directory of the user _shiny_**

`shiny@virtual-server:~$ sudo logout`

`admin@virtual-server:~$ sudo mkdir /srv/shiny-server/`

`admin@virtual-server:~$ sudo usermod -d /srv/shiny-server/ -m shiny`

`admin@virtual-server:~$ sudo chown -R shiny:shiny /srv/shiny-server`

`su - shiny`

<br/>
	
**6. Install RStudio Server** 
Follow instructions on https://posit.co/download/rstudio-server/ 
The steps below are from December 2024.
`shiny@virtual-server:~$ sudo apt-get install gdebi-core`

`shiny@virtual-server:~$ wget https://download2.rstudio.org/server/jammy/amd64/rstudio-server-2024.09.1-394-amd64.deb`

`shiny@virtual-server:~$ sudo gdebi rstudio-server-2024.09.1-394-amd64.deb`

`shiny@virtual-server:~$ sudo systemctl start rstudio-server`

`shiny@virtual-server:~$ sudo systemctl enable rstudio-server`

<br/>

**7. Install the Shiny server** 
Follow instructions on https://posit.co/download/shiny-server/ 
The steps below are from December 2024.
`shiny@virtual-server:~$ sudo su - -c "R -e \"install.packages('shiny', repos='https://cran.rstudio.com/')\"" `

`shiny@virtual-server:~$ wget https://download3.rstudio.org/ubuntu-18.04/x86_64/shiny-server-1.5.22.1017-amd64.deb`

`shiny@virtual-server:~$ sudo gdebi shiny-server-1.5.22.1017-amd64.deb`

<br/>

**8. Create a requal project in RStudio server**
Open the RStudio server at your server address, port 8787, for example http://my-server-ip-address:8787/
Log in as user _shiny_
New project -> new directory _requal_ -> New project -> Create directory.

<br/>

**9. Install Requal**
In RStudio, install the renv package
`install.packages("renv")`

Then enter following commands

`renv::install("RE-QDA/requal")`

`renv::snapshot(prompt = FALSE)`

`renv::isolate()`

<br/>

**10. Create app.R and init.R files in the project**
In the _requal_ directory, create an R file named int.R. Fill the file with the following information:
```
credentials <- data.frame(
  user = "admin",
  password = "admin_password",
  # password will automatically be hashed
  admin = TRUE,
  user_id = 1,
  user_name = "administrator",
  user_mail = "mail@example.com",
  project_admin = TRUE,
  stringsAsFactors = FALSE
)

# Init the database
shinymanager::create_db(
  credentials_data = credentials,
  sqlite_path = "requal_users.sqlite",
  passphrase = "shiny_user_password"
)
```
Click on Source to run the script. 
You should see the requal_users.sqlite file created in the working directory.
<br/>
Then create an R file named app.R. Fill the file with the following information:

```
requal::run_app(
  mode = "server",
  dbname = "requal",
  dbusername = "shiny",
  dbpassword = "password_to_your_database",
  credentials_path = "requal_users.sqlite",
  credentials_pass = "shiny_user_password"
)
```
Save the file and type in the terminal
`shiny@virtual-server:~$ sudo systemctl restart shiny-server`
<br/>
**11. In your browser, type http://my-server-ip-address:3838/requal/**

If Requal starts up fine, type in the terminal
`shiny@virtual-server:~$ sudo deluser shiny sudo`
